tinker_cookbook/renderers.py

```diff
@@ -1062,32 +1276,173 @@ def _end_message_token(self) -> int:
+        for match in re.finditer(
+            r"<｜tool▁call▁begin｜>(\w+)(?:<｜tool▁sep｜>|\s*)(.*?)<｜tool▁call▁end｜>",
+            calls_match.group(1),
+            re.DOTALL,
+        ):
+            raw_text = match.group(0)
+            func_name, args_str = match.group(1), match.group(2).strip()
+
+            # Strip ```json...``` wrapper if present (model sometimes outputs this)
+            if args_str.startswith("```"):
+                # Remove opening ``` (with optional language tag) and closing ```
+                args_str = re.sub(r"^```(?:json)?\s*", "", args_str)
+                args_str = re.sub(r"\s*```$", "", args_str).strip()
+
+            try:
+                json.loads(args_str)
+                tool_calls.append(
+                    ToolCall(function=ToolCall.FunctionBody(name=func_name, arguments=args_str))
+                )
+            except json.JSONDecodeError as e:
+                unparsed_tool_calls.append(
+                    UnparsedToolCall(raw_text=raw_text, error=f"Invalid JSON: {e}")
+                )
+
+        return tool_calls, unparsed_tool_calls
+
     def parse_response(self, response: list[int]) -> tuple[Message, bool]:
-        return parse_response_for_stop_token(response, self.tokenizer, self._end_message_token)
+        assistant_message, parse_success = parse_response_for_stop_token(
+            response, self.tokenizer, self._end_message_token
+        )
+        if not parse_success:
+            return assistant_message, False
+
+        assert isinstance(assistant_message["content"], str)
+        content = assistant_message["content"]
+
+        # Parse tool calls
+        tool_calls, unparsed_tool_calls = self._parse_deepseek_tool_calls(content)
+        if tool_calls:
+            assistant_message["tool_calls"] = tool_calls
+        if unparsed_tool_calls:
+            assistant_message["unparsed_tool_calls"] = unparsed_tool_calls
+
+        # Strip tool calls section from content (both parsed and unparsed)
+        if tool_calls or unparsed_tool_calls:
+            content = re.sub(
+                r"\s*<｜tool▁calls▁begin｜>.*?<｜tool▁calls▁end｜>",
+                "",
+                content,
+                flags=re.DOTALL,
+            )
+            assistant_message["content"] = content.strip()
+
+        return assistant_message, True
+
+    def create_conversation_prefix_with_tools(
+        self, tools: list[ToolSpec], system_prompt: str = ""
+    ) -> list[Message]:
+        """Create system/user message with DeepSeek V3 tool specifications.
+
+        DeepSeek V3 uses JSON function definitions in the system prompt.
+        Tool calls use special tokens with JSON in code blocks.
+
+        Note: DeepSeek V3 doesn't support system role natively. Returns user role
+        unless system_role_as_user is True (in which case system messages are
+        automatically converted to user role by render_message).
```

Comment:
I think Deepseek does accept having an initial system message (which doesn't have the system header). Can you double check the chat template for this?